wpilibj/src/main/java/edu/wpi/first/wpilibj/smartdashboard/MechanismJoint2d.java

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+package edu.wpi.first.wpilibj.smartdashboard;
+
+import edu.wpi.first.networktables.NetworkTable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+
+/**
+ * Joint on a Mechanism2d.
+ *
+ * <p>Use {@link #getLigament(String)} to create a ligament based on this joint.
+ *
+ * @see Mechanism2d
+ */
+public class MechanismJoint2d {
+  /** NT path relative to base table. */
+  private final String m_path;
+  /** Keys are the NT paths relative to this joint. */
+  private final Map<String, MechanismLigament2d> m_ligaments = new HashMap<>();
+
+  private final double[] m_pos;
+  private NetworkTable m_table;
+
+  /**
+   * Package-private constructor for root joints.
+   *
+   * @param path NT path, relative to base table.
+   * @param x x coordinate of root (provide only when constructing a root node)
+   * @param y y coordinate of root (provide only when constructing a root node)
+   */
+  MechanismJoint2d(String path, double x, double y) {
```

Comment:
I don't think that there's a good way of doing that:
- a separate map wouldn't be practical to synchronize; and it duplicates lookup operations
- a Root class would be a new mess to deal with, and neither performant nor practical
- the current way isn't very good either- root position can't be changed without violating access modifiers. 

I must say that I'm not sure of the need for the Joint class - it has very little functionality, and if we replace it with a Root class it'll fix the above problem. The `getLigament()` method can be moved to the Ligament class, which will make traversal cleaner without the `getEnd()` calls.

Reply:
I see where you're coming from, but I really want to have support for more than just ligaments (lines).  It would be really nice to e.g. be able to draw a rectangle instead of a line, or draw a circle at the end of a ligament.  Maybe those can all be methods on ligament but it doesn't feel as clean as having them be methods on Joint.

Reply:
But let's give that idea a shot.

Reply:
Another alternative approach might be to do an "addChild" with an interface and different concrete types, rather than doing a factory function for each object type.  In other words there would be a MechanismObject interface, and then Ligament, Circle, etc would be variants of that object that could be attached in various ways to a MechanismObject or Root parent?

Reply:
I think that the interface idea might be best, and would be the easiest to extend to more object types. I'll try to scribble something down soon.

Reply:
The main downside is that doing that looks a little uglier in C++ (as you have to be doing make_shared or make_unique so you're passing pointers instead of values), but it should be reasonably clean in Java.

Reply:
I'm not sure how to divide the responsibilities with the minimum code duplication. I'm considering using an abstract class instead, but it's becoming another mess but significantly better with `protected` members and fields. My current suggestion to API layout:
```yaml
class Mechanism2d implements Sendable:
  public Mechanism2d(double, double)
  public MechanismRoot2d getRoot(String, double, double)

abstract class MechanismObject2d:
  public MechanismObject2d(String)
  public final MechanismObject2d append(MechanismObject2d)  //  returns `this` for chaining calls?
  private final update(NetworkTable)
  protected abstract void updateEntries(NetworkTable)
  public final String getName()

class MechanismRoot2d extends MechanismObject2d:
  package MechanismRoot2d(String, double, double) // to avoid free creation of roots
  public void setPosition(double, double)

class MechanismLigament2d extends MechanismObject2d:
  public MechanismLigament2d(String)
  public void setAngle(Rotation2d)  // angle relative to parent
  public void setAngle(double)  // angle relative to parent
  public double getAngle()
  public void setLength(double)
  public double getLength()
  public void setColor(Color8Bit)
  public void setWeight(double)  // line thickness

// other object types are similar
```

I don't know how this will work in C++, and I'll experiment with Java later. One problem is that it will majorly break if a root is appended, or an object is added twice. This will also have no way of retrieving an existing object.

Thoughts?

Reply:
That's pretty similar to the direction I was thinking.  I was thinking about what "append" returns--should it return this, or the passed-in value?  I think I'm landing in the same spot as you, as returning `this` is more common, and if you want to chain in a nested way, you can simply nest the append calls (e.g. `new Ligament("foo").append(new Ligament("bar").append(new Ligament("baz")))`).

I'm okay with not having the ability to retrieve an existing object--when the user is building out their ligaments they can simply save a reference to the objects as they create them; that will probably feel more natural anyway.

We can at least detect duplicates and either throw an exception or ignore+print an error (probably the latter).

Regarding appending root, we can change the Root to simply not implement the MechanismObject2d interface (it would have basically the same functions, but not implement the interface, so it couldn't be passed into another append() function).

Reply:
I meant to return the passed object - returning `this` wouldn't help much. The idea is to have something like this possible:
```java
Ligament ligB = ligA.append(new Ligament("b");
```